### PR TITLE
support postgresql create index without index name

### LINKED ISSUE #10552: PostgreSQL Proxy create unnamed index cause NullPointerException
### Which version of ShardingSphere did you use?
master

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
Proxy

### Expected behavior
Create index

### Actual behavior

```
sharding_db=> create index on t_order (v);
ERROR:  java.lang.NullPointerException
```

This issue is related to #10524.

### PR DESCRIPTION
**Pull request by strongduanmu**
Fixes #10552.

Changes proposed in this pull request:
- support postgresql create index without index name
- add rewrite test case

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/ddl/CreateIndexStatementContext.java [hunk 1]
```diff
@@ -49,6 +56,10 @@ public CreateIndexStatementContext(final CreateIndexStatement sqlStatement) {
     
     @Override
     public Collection<IndexSegment> getIndexes() {
-        return null == getSqlStatement().getIndex() ? Collections.emptyList() : Collections.singletonList(getSqlStatement().getIndex());
+        Optional<Integer> startIndex = CreateIndexStatementHandler.getGeneratedIndexStartIndex(getSqlStatement());
+        String generatedLogicIndexName = IndexMetaDataUtil.getGeneratedLogicIndexName(getSqlStatement().getColumns());
+        Collection<IndexSegment> generatedIndexSegments = startIndex.map(index -> Collections.singletonList(new IndexSegment(index, index,
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/ddl/CreateIndexStatementContext.java:61]: if (null != getSqlStatement().getIndex()) {
return
}
Collection<IndexSegment> generatedIndexSegments =xxx;
return?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/ddl/impl/ShardingCreateIndexStatementValidator.java [hunk 2]
```diff
@@ -38,7 +39,9 @@ public void preValidate(final ShardingRule shardingRule, final SQLStatementConte
                             final List<Object> parameters, final ShardingSphereSchema schema) {
         validateTableExist(schema, Collections.singletonList(sqlStatementContext.getSqlStatement().getTable()));
         String tableName = sqlStatementContext.getSqlStatement().getTable().getTableName().getIdentifier().getValue();
-        String indexName = sqlStatementContext.getSqlStatement().getIndex().getIdentifier().getValue();
+        String indexName = null != sqlStatementContext.getSqlStatement().getIndex() 
+                ? sqlStatementContext.getSqlStatement().getIndex().getIdentifier().getValue()
+                : IndexMetaDataUtil.getGeneratedLogicIndexName(sqlStatementContext.getSqlStatement().getColumns());
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/ddl/impl/ShardingCreateIndexStatementValidator.java:44]: How about `sqlStatementContext.getIndexName()`?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rewrite/token/pojo/IndexToken.java [hunk 3]
```diff
@@ -62,7 +63,12 @@ public IndexToken(final int startIndex, final int stopIndex, final IdentifierVal
     
     @Override
     public String toString(final RouteUnit routeUnit) {
-        return identifier.getQuoteCharacter().wrap(getIndexValue(routeUnit));
+        String quotedIndexName = identifier.getQuoteCharacter().wrap(getIndexValue(routeUnit));
+        return isGeneratedIndex() ? " " + quotedIndexName + " " : quotedIndexName;
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rewrite/token/pojo/IndexToken.java:67]: Will it be better to move blank space to the outer?

**Review by tristaZero — CHANGES_REQUESTED** [review]

**Comment by kimmking** [comment]
fix ut in windows first

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/ddl/CreateIndexStatementContext.java (hunk 1 above)
Comment (strongduanmu) [reply shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/ddl/CreateIndexStatementContext.java]: > if (null != getSqlStatement().getIndex()) {
> return
> }
> Collection generatedIndexSegments =xxx;
> return?

@tristaZero Thanks for your suggestion, I have optimized it.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/ddl/impl/ShardingCreateIndexStatementValidator.java (hunk 2 above)
Comment (strongduanmu) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/ddl/impl/ShardingCreateIndexStatementValidator.java]: > How about `sqlStatementContext.getIndexName()`?

@tristaZero Great idea, I have extracted the logic into the getIndexes method in sqlStatementContext.

**Comment by strongduanmu** [comment]
> fix ut in windows first

@kimmking Thanks for your reminder, I have fixed it.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rewrite/token/pojo/IndexToken.java (hunk 3 above)
Comment (strongduanmu) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rewrite/token/pojo/IndexToken.java:67]: > Will it be better to move blank space to the outer?

@tristaZero Extracting to the outside means that we need to pass more parameters in RouteSQLBuilder. In addition, I refer to other tokens, and similar string splicing also exists.

```java
// ShardingInsertValuesToken
@Override
public String toString(final RouteUnit routeUnit) {
    StringBuilder result = new StringBuilder();
    appendInsertValue(routeUnit, result);
    result.delete(result.length() - 2, result.length());
    return result.toString();
}

private void appendInsertValue(final RouteUnit routeUnit, final StringBuilder stringBuilder) {
    for (InsertValue each : getInsertValues()) {
        if (isAppend(routeUnit, (ShardingInsertValue) each)) {
            stringBuilder.append(each).append(", ");
        }
    }
}

// OrderByToken
@Override
public String toString() {
    StringBuilder result = new StringBuilder();
    result.append(" ORDER BY ");
    for (int i = 0; i < columnLabels.size(); i++) {
        if (0 == i) {
            result.append(columnLabels.get(0)).append(" ").append(orderDirections.get(i).name());
        } else {
            result.append(",").append(columnLabels.get(i)).append(" ").append(orderDirections.get(i).name());
        }
    }
    result.append(" ");
    return result.toString();
}
```

**Review by tristaZero — APPROVED** [review]

**Pull request merged**